File: src/python/api_client.py
"""
API 客户端

与 ai-werewolf 服务器通信
"""
import json
import time
import requests
from typing import Dict, Any


class ApiClient:
    """API 客户端类"""

    def __init__(self, config: Dict[str, Any]):
        """
        初始化 API 客户端

        Args:
            config: 配置字典
                - apiBaseUrl: API 基础地址
                - gameToken: 游戏访问 token
        """
        self.api_base_url = config.get("apiBaseUrl")
        self.game_token = config.get("gameToken")

    def get_game_status(self, game_id: str) -> Dict[str, Any]:
        """
        获取游戏状态

        Args:
            game_id: 游戏 ID

        Returns:
            游戏状态响应
        """
        url = f"{self.api_base_url}/api/player-agent/game/{game_id}/status"

        print("[API] 📤 发送请求:")
        print("  URL:", url)
        print("  Method: GET")
        print("  Headers:", {
            "Authorization": f"Bearer {self.game_token[:20]}...",
            "Content-Type": "application/json",
        })

        start_time = time.time()
        try:
            response = requests.get(
                url,
                headers={
                    "Authorization": f"Bearer {self.game_token}",
                    "Content-Type": "application/json",
                },
                timeout=10,
            )

            elapsed = int((time.time() - start_time) * 1000)

            if not response.ok:
                print(
                    f"[API] ❌ 响应失败: {response.status_code} {response.reason} ({elapsed}ms)"
                )
                raise Exception(f"HTTP {response.status_code}: {response.reason}")

            data = response.json()
            print(f"[API] ✅ 响应成功: {response.status_code} ({elapsed}ms)")
            print("[API] 📥 响应数据:", json.dumps(data, ensure_ascii=False, indent=2))

            return data
        except requests.exceptions.RequestException as e:
            elapsed = int((time.time() - start_time) * 1000)
            print(f"[API] ❌ 请求失败 ({elapsed}ms): {str(e)}")
            raise

    def send_ready(self, game_id: str) -> Dict[str, Any]:
        """
        发送准备就绪信号

        Args:
            game_id: 游戏 ID

        Returns:
            准备响应
        """
        url = f"{self.api_base_url}/api/player-agent/game/{game_id}/ready"

        print("[API] 📤 发送准备请求:")
        print("  URL:", url)
        print("  Method: POST")
        print("  Headers:", {
            "Authorization": f"Bearer {self.game_token[:20]}...",
            "Content-Type": "application/json",
        })

        start_time = time.time()
        try:
            response = requests.post(
                url,
                headers={
                    "Authorization": f"Bearer {self.game_token}",
                    "Content-Type": "application/json",
                },
                timeout=10,
            )

            elapsed = int((time.time() - start_time) * 1000)

            if not response.ok:
                try:
                    error_data = response.json()
                except:
                    error_data = {}
                print(
                    f"[API] ❌ 响应失败: {response.status_code} {response.reason} ({elapsed}ms)"
                )
                print("[API] 错误详情:", json.dumps(error_data, ensure_ascii=False, indent=2))
                error_msg = error_data.get("error", {}).get("message", response.reason)
                raise Exception(f"HTTP {response.status_code}: {error_msg}")

            data = response.json()
            print(f"[API] ✅ 响应成功: {response.status_code} ({elapsed}ms)")
            print("[API] 📥 响应数据:", json.dumps(data, ensure_ascii=False, indent=2))

            return data
        except requests.exceptions.RequestException as e:
            elapsed = int((time.time() - start_time) * 1000)
            print(f"[API] ❌ 请求失败 ({elapsed}ms): {str(e)}")
            raise

    def submit_action(self, game_id: str, action: Dict[str, Any]) -> Dict[str, Any]:
        """
        提交游戏行动

        Args:
            game_id: 游戏 ID
            action: 行动数据

        Returns:
            行动响应
        """
        url = f"{self.api_base_url}/api/player-agent/game/{game_id}/action"

        print("[API] 📤 发送请求:")
        print("  URL:", url)
        print("  Method: POST")
        print("  Headers:", {
            "Authorization": f"Bearer {self.game_token[:20]}...",
            "Content-Type": "application/json",
        })
        print("  Body:", json.dumps(action, ensure_ascii=False, indent=2))

        start_time = time.time()
        try:
            response = requests.post(
                url,
                headers={
                    "Authorization": f"Bearer {self.game_token}",
                    "Content-Type": "application/json",
                },
                json=action,
                timeout=10,
            )

            elapsed = int((time.time() - start_time) * 1000)

            if not response.ok:
                try:
                    error_data = response.json()
                except:
                    error_data = {}
                print(
                    f"[API] ❌ 响应失败: {response.status_code} {response.reason} ({elapsed}ms)"
                )
                print("[API] 错误详情:", json.dumps(error_data, ensure_ascii=False, indent=2))
                error_msg = error_data.get("error", {}).get("message", response.reason)
                raise Exception(f"HTTP {response.status_code}: {error_msg}")

            data = response.json()
            print(f"[API] ✅ 响应成功: {response.status_code} ({elapsed}ms)")
            print("[API] 📥 响应数据:", json.dumps(data, ensure_ascii=False, indent=2))

            return data
        except requests.exceptions.RequestException as e:
            elapsed = int((time.time() - start_time) * 1000)
            print(f"[API] ❌ 请求失败 ({elapsed}ms): {str(e)}")
            raise

File: src/js/api-client.js
/**
 * API 客户端
 *
 * 与 ai-werewolf 服务器通信
 */

/**
 * API 客户端类
 */
export class ApiClient {
  /**
   * @param {Object} config
   * @param {string} config.apiBaseUrl - API 基础地址
   * @param {string} config.gameToken - 游戏访问 token
   */
  constructor(config) {
    this.apiBaseUrl = config.apiBaseUrl;
    this.gameToken = config.gameToken;
  }

  /**
   * 获取游戏状态
   * @param {string} gameId - 游戏 ID
   * @returns {Promise<import('./types.js').GameStatusResponse>}
   */
  async getGameStatus(gameId) {
    const url = `${this.apiBaseUrl}/api/player-agent/game/${gameId}/status`;

    console.log("[API] 📤 发送请求:");
    console.log("  URL:", url);
    console.log("  Method: GET");
    console.log("  Headers:", {
      Authorization: `Bearer ${this.gameToken.substring(0, 20)}...`,
      "Content-Type": "application/json",
    });

    const startTime = Date.now();
    const response = await fetch(url, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${this.gameToken}`,
        "Content-Type": "application/json",
      },
    });

    const elapsed = Date.now() - startTime;

    if (!response.ok) {
      console.log(
        `[API] ❌ 响应失败: ${response.status} ${response.statusText} (${elapsed}ms)`
      );
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const data = await response.json();
    console.log(`[API] ✅ 响应成功: ${response.status} (${elapsed}ms)`);
    console.log("[API] 📥 响应数据:", JSON.stringify(data, null, 2));

    return data;
  }

  /**
   * 发送准备就绪信号
   * @param {string} gameId - 游戏 ID
   * @returns {Promise<import('./types.js').ReadyResponse>}
   */
  async sendReady(gameId) {
    const url = `${this.apiBaseUrl}/api/player-agent/game/${gameId}/ready`;

    console.log("[API] 📤 发送准备请求:");
    console.log("  URL:", url);
    console.log("  Method: POST");
    console.log("  Headers:", {
      Authorization: `Bearer ${this.gameToken.substring(0, 20)}...`,
      "Content-Type": "application/json",
    });

    const startTime = Date.now();
    const response = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.gameToken}`,
        "Content-Type": "application/json",
      },
    });

    const elapsed = Date.now() - startTime;

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      console.log(
        `[API] ❌ 响应失败: ${response.status} ${response.statusText} (${elapsed}ms)`
      );
      console.log("[API] 错误详情:", JSON.stringify(errorData, null, 2));
      throw new Error(
        `HTTP ${response.status}: ${
          errorData.error?.message || response.statusText
        }`
      );
    }

    const data = await response.json();
    console.log(`[API] ✅ 响应成功: ${response.status} (${elapsed}ms)`);
    console.log("[API] 📥 响应数据:", JSON.stringify(data, null, 2));

    return data;
  }

  /**
   * 提交游戏行动
   * @param {string} gameId - 游戏 ID
   * @param {Object} action - 行动数据
   * @returns {Promise<import('./types.js').ActionResponse>}
   */
  async submitAction(gameId, action) {
    const url = `${this.apiBaseUrl}/api/player-agent/game/${gameId}/action`;

    console.log("[API] 📤 发送请求:");
    console.log("  URL:", url);
    console.log("  Method: POST");
    console.log("  Headers:", {
      Authorization: `Bearer ${this.gameToken.substring(0, 20)}...`,
      "Content-Type": "application/json",
    });
    console.log("  Body:", JSON.stringify(action, null, 2));

    const startTime = Date.now();
    const response = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.gameToken}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(action),
    });

    const elapsed = Date.now() - startTime;

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      console.log(
        `[API] ❌ 响应失败: ${response.status} ${response.statusText} (${elapsed}ms)`
      );
      console.log("[API] 错误详情:", JSON.stringify(errorData, null, 2));
      throw new Error(
        `HTTP ${response.status}: ${
          errorData.error?.message || response.statusText
        }`
      );
    }

    const data = await response.json();
    console.log(`[API] ✅ 响应成功: ${response.status} (${elapsed}ms)`);
    console.log("[API] 📥 响应数据:", JSON.stringify(data, null, 2));

    return data;
  }
}
